### PR TITLE
Save favourite documents & collection and default workspace to backend #LMR-880

### PR DESCRIPTION
**Pull request by kubedo8**

### CODE REVIEW

File: src/app/core/guards/page-not-found.guard.ts [hunk 1]
```diff
@@ -19,30 +19,49 @@
...
-    if (w === 'w' && organizationCode && projectCode) {
-      this.router.navigate(['w', organizationCode, projectCode, 'view', 'search']);
-    } else {
-      this.router.navigate(['workspace']);
-    }
+    return this.getDefaultWorkspace().pipe(
+      map(workspace => {
+        const hasWorkspace = workspace && workspace.organizationCode && workspace.projectCode;
+        if (w === 'w' && organizationCode && projectCode && hasWorkspace) {
+          this.router.navigate(['w', workspace.organizationCode, workspace.projectCode, 'view', 'search']);
```
Comment (livthomas) [inline src/app/core/guards/page-not-found.guard.ts]: This behavior does not make sense to me. You check if both the default workspace and workspace from the URL is set and then redirect it to the default one.

I think it should make redirects in this order:

1. If workspace is defined in the URL, navigate to it.
2. If not and default workspace is defined, navigate to it.
3. Otherwise, navigate to workspace selection page.

What do you think?

File: src/app/core/guards/view/view-exist.guard.ts [hunk 2]
```diff
@@ -72,13 +72,15 @@ export class ViewExistGuard implements CanActivate {
...
     this.notificationsService.error(message);
+    console.log('view not found');
     return false;
   }
 
   private redirectToCorrectPath(view: ViewModel, workspace: Workspace) {
     const perspective = view.perspective ? view.perspective : Perspective.Search;
     const query = QueryConverter.toString(view.query);
     this.router.navigate(['w', workspace.organizationCode, workspace.projectCode, 'view', {vc: view.code}, perspective,], {queryParams: {query}});
+    console.log('redirect to correct path');
```
Comment (livthomas) [inline src/app/core/guards/view/view-exist.guard.ts]: Forgotten `console.log` calls.

File: src/app/core/store/collections/collections.effects.ts [hunk 3]
```diff
@@ -115,10 +114,10 @@ export class CollectionsEffects {
     withLatestFrom(this.store$.select(selectOrganizationByWorkspace)),
     map(([action, organization]) => {
       if (action.payload.error instanceof HttpErrorResponse && action.payload.error.status == 402) {
-        const title = this.i18n({id: 'serviceLimits.trial', value: 'Free Service'});
+        const title = this.i18n({id: 'serviceLimits.trial', value: 'Trial Service'});
```
Comment (livthomas) [inline src/app/core/store/collections/collections.effects.ts]: Are these changes intentional or a result of a bad merge?

File: src/app/core/store/collections/collections.action.ts [hunk 4]
```diff
@@ -219,7 +219,7 @@ export namespace CollectionsAction {
   export class AddFavoriteFailure implements Action {
     public readonly type = CollectionsActionType.ADD_FAVORITE_FAILURE;
 
-    public constructor(public payload: { error: any }) {
+    public constructor(public payload: { collectionId: string, error: any }) {
```
Comment (livthomas) [inline src/app/core/store/collections/collections.action.ts]: Why do we still need these actions? Is it not updated with the collection?

File: src/app/core/store/documents/documents.action.ts [hunk 5]
```diff
@@ -31,13 +31,20 @@ export enum DocumentsActionType {
...
   UPDATE_DATA = '[Documents] Update Data',
   PATCH_DATA = '[Documents] Patch Data',
 
+  ADD_FAVORITE = '[Documents] Add Favorite',
+  ADD_FAVORITE_SUCCESS = '[Documents] Add Favorite :: Success',
+  ADD_FAVORITE_FAILURE = '[Documents] Add Favorite :: Failure',
+
+  REMOVE_FAVORITE = '[Documents] Remove Favorite',
+  REMOVE_FAVORITE_SUCCESS = '[Documents] Remove Favorite :: Success',
+  REMOVE_FAVORITE_FAILURE = '[Documents] Remove Favorite :: Failure',
```
Comment (livthomas) [inline src/app/core/store/documents/documents.action.ts]: Why do we have new actions for these? We should not add a new action for every field in document metadata.

File: src/app/core/rest/global.service.ts [hunk 6]
```diff
@@ -34,6 +35,10 @@ export class GlobalService {
     return this.httpClient.get<User>(`${this.apiPrefix()}/currentUser`);
   }
 
+  public saveDefaultWorkspace(defaultWorkspace: DefaultWorkspace): Observable<any> {
+    return this.httpClient.put(`${this.apiPrefix()}/workspace`, defaultWorkspace);
```
Comment (livthomas) [inline src/app/core/rest/global.service.ts]: Why is it not defined in the user service if it modifies a user?

File: src/app/core/store/users/users.effects.ts [hunk 7]
```diff
@@ -152,6 +154,18 @@ export class UsersEffects {
...
   );
 
+  @Effect()
+  public saveDefaultWorkspace$ = this.actions$.pipe(
+    ofType<UsersAction.SaveDefaultWorkspace>(UsersActionType.SAVE_DEFAULT_WORKSPACE),
+    concatMap(action => {
+      const defaultWorkspaceDto = DefaultWorkspaceConverter.toDto(action.payload.defaultWorkspace);
+      return this.globalService.saveDefaultWorkspace(defaultWorkspaceDto).pipe(
+        concatMap(() => Observable.of()),
+        catchError(() => Observable.of())
```
Comment (livthomas) [inline src/app/core/store/users/users.effects.ts]: Why do we need `concatMap` here?

File: src/app/shared/post-it-collections/post-it-collections.module.ts [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    PostItCollectionComponent,
+    PostItCollectionNameComponent,
+    PostItCollectionAddButtonComponent,
+    PostItCollectionImportButtonComponent,
+  ],
+  exports: [
+    PostItCollectionsComponent
+  ]
+})
+export class PostItCollectionsModule {
```
Comment (livthomas) [inline src/app/shared/post-it-collections/post-it-collections.module.ts]: +1 on creating a new module for these components.

File: src/app/core/dto/default-workspace.ts [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+export interface DefaultWorkspace {
+
+  organizationCode?: string;
+  organizationId: string;
+  projectCode?: string;
+  projectId: string;
```
Comment (livthomas) [inline src/app/core/dto/default-workspace.ts]: Why do we need both codes and IDs?

File: src/app/core/guards/page-not-found.guard.ts [hunk 10]
```diff
@@ -19,30 +19,49 @@
...
+    );
 
-    const message = this.i18n({id: 'page.not.found', value: 'Page not found'});
-    this.notificationService.error(message);
+  }
 
-    return false;
+  private getDefaultWorkspace(): Observable<DefaultWorkspaceModel> {
+    return this.store.select(selectCurrentUser).pipe(
+      map(user => user.defaultWorkspace)
```
[the remark this answers is not in the document]
Comment (kubedo8) [reply src/app/core/guards/page-not-found.guard.ts]: Ok

File: src/app/core/guards/page-not-found.guard.ts (hunk 1 above)
Comment (kubedo8) [reply src/app/core/guards/page-not-found.guard.ts]: Ok, that make sense

File: src/app/core/guards/view/view-exist.guard.ts (hunk 2 above)
Comment (kubedo8) [reply src/app/core/guards/view/view-exist.guard.ts]: Thanks

File: src/app/core/store/collections/collections.effects.ts (hunk 3 above)
Comment (kubedo8) [reply src/app/core/store/collections/collections.effects.ts]: Bad merge :)

File: src/app/core/store/collections/collections.action.ts (hunk 4 above)
Comment (kubedo8) [reply src/app/core/store/collections/collections.action.ts]: Of course not. We store information about favorite collections and documents in separate collections, because it's individual for each user.

File: src/app/core/store/documents/documents.action.ts (hunk 5 above)
Comment (kubedo8) [reply src/app/core/store/documents/documents.action.ts]: It's not metadata. See comment above.

File: src/app/core/rest/global.service.ts (hunk 6 above)
Comment (kubedo8) [reply src/app/core/rest/global.service.ts]: Because `UserService` has `organizationId` in path and there we don't need it.

File: src/app/core/store/users/users.effects.ts (hunk 7 above)
Comment (kubedo8) [reply src/app/core/store/users/users.effects.ts]: Maybe it's useless..

File: src/app/core/dto/default-workspace.ts (hunk 9 above)
Comment (kubedo8) [reply src/app/core/dto/default-workspace.ts]: Because we are sending `ids` but receive both. Codes are computed before send, so when organization and project is changed, we send fresh code for navigate to correct url.

File: src/app/core/store/users/users.effects.ts (hunk 7 above)
Comment (kubedo8) [reply src/app/core/store/users/users.effects.ts]: When I removed it, I got an error.

File: src/app/core/store/collections/collections.action.ts (hunk 4 above)
Comment (livthomas) [reply src/app/core/store/collections/collections.action.ts]: Oh, I see. But we return this information from the back-end inside collection/document object, right?

File: src/app/core/rest/global.service.ts (hunk 6 above)
Comment (livthomas) [reply src/app/core/rest/global.service.ts]: I am thinking if this is the right approach. We already have two REST methods which work with users but cannot be in user service because they do not need `organizationId`. Maybe it would be worth changing user service to have `organizationId` only on those methods which need it. (Just an idea.)

File: src/app/core/store/collections/collections.action.ts (hunk 4 above)
Comment (kubedo8) [reply src/app/core/store/collections/collections.action.ts]: Yes :)
